Stratis.Bitcoin/FullNode.cs

```diff
@@ -155,18 +157,125 @@ public void Start()
 
 			// start all the features defined
 			this.StartFeatures();
 
 			ConnectionManager.Start();
-			_IsStarted.Set();
+			_isStarted.Set();
 
 			this.StartPeriodicLog();
 		}
 
-		private BlockStore.ChainBehavior.ChainState _ChainBehaviorState;
+	    public void Run()
+	    {
+	        RunAsync().GetAwaiter().GetResult();
+	    }
+
+        public async Task RunAsync()
+	    {
+	        var done = new ManualResetEventSlim(false);
+	        using (var cts = new CancellationTokenSource())
+	        {
+	            Action shutdown = () =>
+	            {
+	                if (!cts.IsCancellationRequested)
+	                {
+	                    Console.WriteLine("Application is shutting down...");
+	                    try
+	                    {
+	                        cts.Cancel();
+	                    }
+	                    catch (ObjectDisposedException)
+	                    {
+	                    }
+	                }
+
+	                done.Wait();
+	            };
+
+                var assemblyLoadContext = AssemblyLoadContext.GetLoadContext(typeof(FullNode).GetTypeInfo().Assembly);
+                assemblyLoadContext.Unloading += context => shutdown();
+                Console.CancelKeyPress += (sender, eventArgs) =>
+	            {
+	                shutdown();
+	                // Don't terminate the process immediately, wait for the Main thread to exit gracefully.
+	                eventArgs.Cancel = true;
+	            };
+
+	            await this.RunAsync(cts.Token, "Application started. Press Ctrl+C to shut down.");
+	            done.Set();
+	        }
+        }
+
+	    public async Task RunAsync(CancellationToken cancellationToken, string shutdownMessage)
+	    {
+	        using (this)
+	        {
+	            await this.StartAsync(cancellationToken);
+
+	            if (!string.IsNullOrEmpty(shutdownMessage))
+	            {
+	                Console.WriteLine(shutdownMessage);
+	            }
+
+                cancellationToken.Register(state =>
+                    {
+                        ((IApplicationLifetime)state).StopApplication();
+                    },
+                    this._applicationLifetime);
+
+                var waitForStop = new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously);
+                this._applicationLifetime.ApplicationStopping.Register(obj =>
+                {
+                    var tcs = (TaskCompletionSource<object>)obj;
+                    tcs.TrySetResult(null);
+                }, waitForStop);
+
+                await waitForStop.Task;
+
+	            await StopAsync();
+	        }
+	    }
+
+	    public async Task StartAsync(CancellationToken cancellationToken)
+	    {
+	        await Task.Run(() => Start(), cancellationToken);
+	    }
+
+	    public async Task StopAsync()
+	    {
+	        await Task.Run(() => Stop(CancellationToken.None));
+	    }
+
+        public void Stop(CancellationToken cancellationToken)
```

Comment:
I am not sure Stop() needs a cancellation token, it will just cancel the global cancellation.